### PR TITLE
Refactor: Start to separate wallet from node

### PR DESCRIPTION
**Pull request by ryanofsky**
This creates an incomplete [`Chain`](https://github.com/ryanofsky/bitcoin/blob/pr/wipc-sep/src/interfaces/chain.h) interface in [`src/interfaces/`](https://github.com/ryanofsky/bitcoin/tree/pr/wipc-sep/src/interfaces) and begins to update wallet code to use it.

#10973 builds on this, changing the wallet to use the new interface to access chain state, instead of using CBlockIndex pointers and global variables like `chainActive`.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
This doesn't compile with clang

**Comment by Empact** [comment]
The error from travis:
```
interfaces/wallet.cpp:109:23: error: calling function 'CheckFinalTx' requires holding mutex 'cs_main' exclusively [-Werror,-Wthread-safety-analysis]
    result.is_final = CheckFinalTx(*wtx.tx);
                      ^
1 error generated.
```

**Comment by ryanofsky** [comment]
Added 2 more small commits d0af21fab95ee453a1fee4242ceb89ca79d568eb -> a10e919db4c9f37f0bb60fd62a232abf8bd3dede ([pr/wchain.1](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.1) -> [pr/wchain.2](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wchain.1...pr/wchain.2)) from #10973 to fix the thread safety warnings (which are harmless but treated like errors).

**Comment by MeshCollider** [comment]
Concept ACK :+1:

**Comment by sipa** [comment]
Is it intentional that the interfaces code uses a different coding style than the rest of the code?

If so, does this need documenting somewhere, and if not, can it be converted?

**Comment by practicalswift** [comment]
This PR does not seem to compile when rebased on `master` :-)

File: src/interfaces/chain.cpp [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        if (height >= 0) {
+            return height;
+        }
+        return nullopt;
+    }
+    Optional<int> getBlockHeight(const uint256& hash) override
+    {
+        auto it = ::mapBlockIndex.find(hash);
+        if (it != ::mapBlockIndex.end() && it->second) {
+            if (::chainActive.Contains(it->second)) {
```
Comment (practicalswift) [inline src/interfaces/chain.cpp]: Collapse into `if` statement above?

**Comment by ryanofsky** [comment]
> I noticed d0af21f adds a bunch of methods to interfaces/chain.cpp. I assume those are moved from somewhere else, but I don't see their counterparts removed in this PR. Why is that?

The new Chain d0af21fab95ee453a1fee4242ceb89ca79d568eb methods aren't moved from someplace else. They are fairly simple utility functions that wrap`CBlockIndex`/`chainActive` APIs so the wallet doesn't access chain state directly.

Also worth noting that d0af21fab95ee453a1fee4242ceb89ca79d568eb is the main commit in this PR. The other commits are basically just mechanical changes leading to up d0af21fab95ee453a1fee4242ceb89ca79d568eb.

**Review by ken2812221 — COMMENTED** [review]
Concept ACK. `msvc-autogen.py` change looks good to me.

**Comment by ryanofsky** [comment]
> would be good to get @ken2812221's feedback on the msvc-autogen.py changes

ken2812221 did previously review those changes: https://github.com/bitcoin/bitcoin/pull/10973#issuecomment-418624533. They were in the first commit of the PR at that time: d487d953c9a10b11693ba8af62f5b54a77ec7686.

**Comment by ryanofsky** [comment]
There were some comments in IRC that the PR description here might need to be improved. Would welcome any suggestions. #10973 has a more complete description of the direction this PR starts to go in.

re: https://github.com/bitcoin/bitcoin/pull/14437#issuecomment-429439919
> Is it intentional that the interfaces code uses a different coding style than the rest of the code?
>
> If so, does this need documenting somewhere, and if not, can it be converted?

I'll create a PR to document the slight style change. It's used in gui code so I adopted it for #10244 and #10102, sticking with it in #10973 and here for consistency, and since I do like it.

**Review by maflcko — COMMENTED** [review]
utACK 0b2cf7c3a8

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -538,6 +537,7 @@ UniValue importwallet(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is currently rescanning. Abort existing rescan or wait.");
     }
 
+    Optional<int> tip_height;
```
Comment (MarcoFalke) [inline src/wallet/rpcdump.cpp]: in commit c4b750f:

Why is this needed?

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -101,7 +101,11 @@ static void WalletTxToJSON(interfaces::Chain& chain, interfaces::Chain::Lock& lo
     {
         entry.pushKV("blockhash", wtx.hashBlock.GetHex());
         entry.pushKV("blockindex", wtx.nIndex);
-        entry.pushKV("blocktime", LookupBlockIndex(wtx.hashBlock)->GetBlockTime());
+        int64_t block_time;
+        if (!chain.findBlock(wtx.hashBlock, nullptr /* CBlock */, &block_time)) {
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: in commit c4b750f:

The name of the argument is `block`, not `CBlock`?

File: src/wallet/rpcwallet.cpp [hunk 4]
```diff
@@ -101,7 +101,11 @@ static void WalletTxToJSON(interfaces::Chain& chain, interfaces::Chain::Lock& lo
...
+            throw std::logic_error("invalid wallet transaction block hash");
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: in commit c4b750f:

It is not immediately clear where this exceptions goes to. Crashes the node?

I'd prefer if we sticked to jsonrpc exceptions in the rpc module, so that broken rpc code would not crash the node.

File: src/wallet/rpcwallet.cpp [hunk 5]
```diff
@@ -1575,9 +1574,9 @@ static UniValue listsinceblock(const JSONRPCRequest& request)
     // when a reorg'd block is requested, we also list any relevant transactions
     // in the blocks of the chain that was detached
     UniValue removed(UniValue::VARR);
-    while (include_removed && paltindex && paltindex != pindex) {
+    while (include_removed && altheight && *altheight > height.value_or(-1)) {
         CBlock block;
-        if (!ReadBlockFromDisk(block, paltindex, Params().GetConsensus())) {
+        if (!pwallet->chain().findBlock(blockId, &block) || block.IsNull()) {
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: in commit c4b750f:

Previously this would take an argument that was modified in each loop run, now the arguments are constant. That doesn't seem right.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -3295,58 +3295,53 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
...
-        stopBlock = pindexStop ? pindexStop : pChainTip;
     }
     else {
         throw JSONRPCError(RPC_MISC_ERROR, "Rescan failed. Potentially corrupted data files.");
     }
     UniValue response(UniValue::VOBJ);
-    response.pushKV("start_height", pindexStart->nHeight);
-    response.pushKV("stop_height", stopBlock->nHeight);
+    response.pushKV("start_height", start_height);
+    response.pushKV("stop_height", stop_height ? *stop_height : tip_height ? *tip_height: UniValue());
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp]: in commit c4b750f:

Couldn't stop_height be initialized to tip_height by default. Also, if tip_height is not available, we might as well exit early, to safe us from plastering the code with question mark operators?

File: src/threadsafety.h [hunk 7]
```diff
@@ -54,4 +54,15 @@
...
 #endif // __GNUC__
 
+// Utility class for indicating to compiler thread analysis that a mutex is
+// locked (when it couldn't be determined otherwise).
+struct SCOPED_LOCKABLE LockAnnotation
+{
+    template <typename Mutex>
+    explicit LockAnnotation(Mutex& mutex) EXCLUSIVE_LOCK_FUNCTION(mutex)
+    {
+    }
```
Comment (MarcoFalke) [inline src/threadsafety.h]: in commit 0b2cf7c3a8edd6feb4a8504675960c9cfcbc613c:

Could add the runtime assert that the lock is taken?

`AssertLockHeld`

**Review by maflcko — COMMENTED** [review]
Concept ACK c4b750f510d1913a6af6549b17bd3d89e211c62c

Last commit is (at least for me) non-trivial to verify its refactoring nature. I am wondering if it could be split up more...

File: src/threadsafety.h (hunk 7 above)
Comment (ryanofsky) [reply src/threadsafety.h]: re: https://github.com/bitcoin/bitcoin/pull/14437#discussion_r230501820

That would make this depend on sync.h and require a bitcoin debug mutex instead of working with any mutex. It would also print a less useful file/line number than a standalone assert. Better imo to have separate locking primitives that each do one job well than to have a combined one that's worse at both jobs.

**Review by ryanofsky — COMMENTED** [review]
Added 3 commits c4b750f510d1913a6af6549b17bd3d89e211c62c -> 4072fbe3adb5289b1ef153952453e809c671a415 ([pr/wchain.3](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.3) -> [pr/wchain.4](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wchain.3...pr/wchain.4)) with suggestions from marco and practicalswift
Added 1 commits 4072fbe3adb5289b1ef153952453e809c671a415 -> e4fe52e0ffb2025a1e2081f044854cdcded01689 ([pr/wchain.4](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.4) -> [pr/wchain.5](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wchain.4...pr/wchain.5)) with suggestions from marco
Rebased e4fe52e0ffb2025a1e2081f044854cdcded01689 -> 081accb875412f38718f2e40ed7bbdf15e6f4ef8 ([pr/wchain.5](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.5) -> [pr/wchain.6](https://github.com/ryanofsky/bitcoin/commits/pr/wchain.6)) due to conflicts with #14350 and #14555. Also dropped main commit 90bbccd2c3d502d6341985b213e8f839ceeedddf  (part of #10973) as suggested by Marco and James to reduce size of PR.

**Comment by MarcoFalke** [comment]
re-utACK 081accb875412f38718f2e40ed7bbdf15e6f4ef8 (Only change since previous utACK 0b2cf7c is rebase and adding the return in a void context)

**Review by ken2812221 — APPROVED** [review]
utACK 081accb875412f38718f2e40ed7bbdf15e6f4ef8

File: src/interfaces/wallet.cpp [hunk 8]
```diff
@@ -456,8 +491,32 @@ class WalletImpl : public Wallet
     CWallet& m_wallet;
 };
 
+class WalletClientImpl : public ChainClient
```
Comment (promag) [inline src/interfaces/wallet.cpp]: IMO this could have a better name, currently this chain client "manages all wallets".

**Review by promag — COMMENTED** [review]
Concept ACK.

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/14437/commits/081accb875412f38718f2e40ed7bbdf15e6f4ef8

I'll do manual testing later today. Summary of reviewed commits:

- [x] 7e2e62cf7 - Add skeleton chain and client classes

  Introduces Chain class which consolidates the node's interface for use by
  clients like GUI and wallet. Also introduces ChainClient which is a class
  that allows the node to manage clients. No methods introduced on either.

- [x] 8db11dd0b - Pass chain and client variables where needed 

  Constructs Chain interface (via InitInterfaces) during
  initialization and passes it to dependent code.

- [x] 79d579f4e - Remove uses of cs_main in wallet code

  Introduces Chain::Lock and uses its acquisition to replace references to
  cs_main in wallet code. Chain::Lock's interface is Chain's, but its use
  implies that we have a lock on the chainstate (i.e. cs_main).

  Eventually this could be replaced with a snapshot of the chainstate to avoid
  blocking more critical chainstate manipulations (and whatever else requires
  cs_main) during wallet operations.

- [x] 081accb87 - Pass chain locked variables where needed

  Makes explicit the dependence of various wallet methods on the Chain
  interface. Passing around `locked_chain` implies cs_main acquisition and so
  per-function lock annotations are removed. In a later PR, we'll actually use
  locked_chain methods to replace direct access to e.g. `chainActive` and
  `mapBlockIndex`.

- [x] ea961c3d7 - Remove direct node->wallet calls in init.cpp

  Replaces wallet-specific initialization with a more general `chain_clients`
  model - allowing for the initialization of non-wallet code which consumes the
  chain in a read-only way.

**Comment by ryanofsky** [comment]
Thanks Marco, James, and Ken for the recent code reviews! James, I think I will steal your breakdown of commits and add it to the PR description.

File: src/init.cpp [hunk 9]
```diff
@@ -1243,7 +1258,11 @@ bool AppInitMain()
     }
 
     // ********************************************************* Step 5: verify wallet database integrity
-    if (!g_wallet_init_interface.Verify()) return false;
+    for (const auto& client : interfaces.chain_clients) {
```
Comment (practicalswift) [inline src/init.cpp]: Could use `std::any_of`? :-)

File: src/init.cpp [hunk 10]
```diff
@@ -1562,7 +1581,11 @@ bool AppInitMain()
     }
 
     // ********************************************************* Step 9: load wallet
-    if (!g_wallet_init_interface.Open()) return false;
+    for (const auto& client : interfaces.chain_clients) {
```
Comment (practicalswift) [inline src/init.cpp]: Could use `std::any_of`? :-)

Comment (MeshCollider) [reply src/init.cpp]: @practicalswift just a meta-note, could you please put all your comments in one review rather than putting them all as separate comments?

Comment (practicalswift) [reply src/init.cpp]: @MeshCollider Sure!

Comment (MarcoFalke) [reply src/init.cpp]: Also, I'd prefer if we didn't leave style feedback at all.

Of course it could use std::any_of, or it could use a while loop, or do while loop, or... It is just a different color up to the author to choose and for everyone else to accept.

Comment (practicalswift) [reply src/init.cpp]: @MarcoFalke I should have made clear that my comment was a "nit" :-)

The comparison you're making isn't fair: no one would suggest changing this to a `while` loop, but striving to use STL algorithms were possible is what C++ Core Guidelines (with editors Bjarne Stroustrup and Herb Sutter) recommends:

> ES.1: Prefer the standard library to other libraries and to “handcrafted code”
>
> Code using a library can be much easier to write than code working directly with language features, much shorter, tend to be of a higher level of abstraction, and the library code is presumably already tested. The ISO C++ Standard Library is among the most widely known and best tested libraries. It is available as part of all C++ Implementations.
> 
> https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es1-prefer-the-standard-library-to-other-libraries-and-to-handcrafted-code

Comment (IntelOSt) [reply src/init.cpp]: cmd.to

**Comment by MeshCollider** [comment]
post-merge-utACK https://github.com/bitcoin/bitcoin/pull/14437/commits/ea961c3d7256c66146b4976ab1293db4a628c0de

This is looking good, thanks Russ! 🎉

File: src/init.cpp (hunk 10 above)
Comment (ryanofsky) [reply src/init.cpp]: > Could use std::any_of? :-)

This I think it would actually have to use std::all_of

Comment (ryanofsky) [reply src/init.cpp]: re: https://github.com/bitcoin/bitcoin/pull/14437#discussion_r232173026

> Could use std::any_of? :-)

This PR is already merged so I won't change this here. But it's an interesting suggestion. I might prefer `all_of` over `any_of` since the bool represents success, and success only happens if _all_ calls succeed.

Note: one drawback to using any/all in a context like this is it makes behavior non-deterministic, because the standard doesn't guarantee what order the calls will run in or whether calls will be skipped due to short-circuiting: https://stackoverflow.com/questions/47692180/is-stdany-of-required-to-follow-short-circuit-logic
